lib/routes/devices-and-sessions.js

```diff
@@ -237,6 +244,99 @@ module.exports = (log, db, config, customs, push, devices) => {
         }
       }
     },
+    {
+      method: 'GET',
+      path: '/account/device/messages',
+      config: {
+        validate: {
+          query: {
+            index: isA.string().alphanum().optional(),
+            limit: isA.number().optional().min(0).max(100),
+          }
+        },
+        auth: {
+          strategy: 'sessionToken'
+        },
+        response: {
+          schema: PUSHBOX_RETRIEVE_SCHEMA
+        }
+      },
+      handler (request, reply) {
+        log.begin('Account.deviceMessages', request)
+
+        const sessionToken = request.auth.credentials
+        const uid = sessionToken.uid
+        const deviceId = sessionToken.deviceId
+        const ip = request.app.clientAddress
+        const query = request.query || {}
+        const {index} = query
+        const limit = query.limit !== undefined ? Math.min(100, query.limit) : 100
+
+        return customs.checkAuthenticated('deviceMessages', ip, uid)
+          .then(() => oAuthMint.getOAuthToken(sessionToken, PUSHBOX_OAUTH_SCOPE_RECV))
+          .then(resp => {
+            const {access_token: oAuthToken} = resp
+            return pushbox.retrieve(uid, deviceId, oAuthToken, limit, index)
+          })
+          .then(reply, reply)
+      }
+    },
+    {
+      method: 'POST',
+      path: '/account/devices/messages',
+      config: {
+        auth: {
+          strategy: 'sessionToken'
+        },
+        validate: {
+          payload: {
+            to: isA.string().required(),
+            topic: isA.string().valid('sendtab').required(),
+            data: isA.string().regex(validators.URL_SAFE_BASE_64).required()
+          }
+        },
+        response: {
+          schema: {}
+        }
+      },
+      handler (request, reply) {
+        log.begin('Account.devicesMessages', request)
+
+        const body = request.payload
+        const sessionToken = request.auth.credentials
+        const uid = sessionToken.uid
+        const ip = request.app.clientAddress
+        const {to, data, topic} = body
+
+        return customs.checkAuthenticated('sendDeviceMessage', ip, uid)
+          .then(() => request.app.devices)
+          .then(devices => {
+            const device = devices.find(d => d.id === to)
+            if (! device) {
+              log.error({
+                op: 'Account.devicesMessages',
+                deviceId: to,
+                uid: uid,
+                error: 'Could not find device'
+              })
+              return
+            }
+            return oAuthMint.getOAuthToken(sessionToken, PUSHBOX_OAUTH_SCOPE_SEND)
+            .then(resp => {
+              const {access_token: oAuthToken} = resp
+              return pushbox.store(uid, device.id, topic, data, oAuthToken)
+            })
+            .then(messageUrl => {
+              const pushReason = MESSAGE_TOPIC_TO_PUSH_REASON[topic]
+              return push.notifyMessageReceived(uid, device, messageUrl, topic, pushReason)
```

Comment:
So just to sanity-check my understanding: We expect *most* messages to be retrieved by the client receiving this push message, extract the URL, and fetching the result straight from pushbox.  They will also fall back to polling the `/account/device/messages` endpoint, but unless they missed a push, they won't find anything there.  Accurate?

Reply:
From [1], it appears that this is not correct, and we expect `GET /account/devices/messages` to be the primary mechanism by which (at least some) browsers retrieve their messages.

[1] https://docs.google.com/document/d/1kOf82sHrIDyPYa_VKzOFiMeIatBqdPg4E2YIIYO1SJ0

Reply:
To add a bit more on that: the URL received by the clients will not be a Pushbox URL, but a FxA `GET /account/devices/messages` url with a `limit=1` queryParam.